x/incentive/keeper/rewards.go

```diff
@@ -0,0 +1,255 @@
+	store := prefix.NewStore(ctx.KVStore(k.key), types.RewardPeriodKeyPrefix)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		var rp types.RewardPeriod
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(iterator.Value(), &rp)
+		if cb(rp) {
+			break
+		}
+	}
+}
+
+// HandleRewardPeriodExpiry deletes expired RewardPeriods from the store and creates a ClaimPeriod in the store for each expired RewardPeriod
+func (k Keeper) HandleRewardPeriodExpiry(ctx sdk.Context, rp types.RewardPeriod) {
+	k.CreateClaimPeriod(ctx, rp.Denom, rp.ClaimEnd, rp.ClaimTimeLock)
+	store := prefix.NewStore(ctx.KVStore(k.key), types.RewardPeriodKeyPrefix)
+	store.Delete(types.GetDenomBytes(rp.Denom))
+	return
+}
+
+// GetRewardPeriod returns the reward period from the store for the input denom and a boolean for if it was found
+func (k Keeper) GetRewardPeriod(ctx sdk.Context, denom string) (types.RewardPeriod, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.RewardPeriodKeyPrefix)
+	bz := store.Get(types.GetDenomBytes(denom))
+	if bz == nil {
+		return types.RewardPeriod{}, false
+	}
+	var rp types.RewardPeriod
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &rp)
+	return rp, true
+}
+
+// SetRewardPeriod sets the reward period in the store for the input deno,
+func (k Keeper) SetRewardPeriod(ctx sdk.Context, rp types.RewardPeriod) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.RewardPeriodKeyPrefix)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(rp)
+	store.Set(types.GetDenomBytes(rp.Denom), bz)
+}
+
+// DeleteRewardPeriod deletes the reward period in the store for the input denom,
+func (k Keeper) DeleteRewardPeriod(ctx sdk.Context, denom string) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.RewardPeriodKeyPrefix)
+	store.Delete(types.GetDenomBytes(denom))
+}
+
+// CreateNewRewardPeriod creates a new reward period from the input reward
+func (k Keeper) CreateNewRewardPeriod(ctx sdk.Context, reward types.Reward) {
+	// reward periods store the amount of rewards payed PER SECOND
+	rewardsPerSecond := sdk.NewDecFromInt(reward.Reward.Amount).Quo(sdk.NewDecFromInt(sdk.NewInt(int64(reward.Duration.Seconds())))).TruncateInt()
+	rewardCoinPerSecond := sdk.NewCoin(reward.Reward.Denom, rewardsPerSecond)
+	rp := types.RewardPeriod{
+		Denom:         reward.Denom,
+		Start:         ctx.BlockTime(),
+		End:           ctx.BlockTime().Add(reward.Duration),
+		Reward:        rewardCoinPerSecond,
+		ClaimEnd:      ctx.BlockTime().Add(reward.Duration).Add(reward.ClaimDuration),
+		ClaimTimeLock: reward.TimeLock,
+	}
+	k.SetRewardPeriod(ctx, rp)
+}
+
+// CreateAndDeleteRewardPeriods creates reward periods for active rewards that don't already have a reward period and deletes reward periods for inactive rewards that currently have a reward period
+func (k Keeper) CreateAndDeleteRewardPeriods(ctx sdk.Context) {
```

Comment:
Correct, after the previous one ends, a new one is pulled based on whatever is in `Params.Rewards`. Assuming params haven't changed it will create an identical reward period with new start and end times.